Lib/fontParts/base/base.py

```diff
@@ -236,125 +307,267 @@ def raiseNotImplementedError(self):
+
+        This is the environment implementation of :meth:`BaseDict.__contains__`.
+
+        :param key: The key to check for. If a :cvar:`BaseDict.keyNormalizer`
+            is set, it will have been applied to the key before checking.
+        :raises NotImplementedError: If the method has not been overridden by a
+            subclass.
+
+        .. important::
+
+            Subclasses must override this method.
+
         """
         self.raiseNotImplementedError()
 
-    has_key = __contains__
+    def __setitem__(self, key: Any, value: Any) -> None:
+        """Set the value for a given key in the object.
+
+        :param key: The key to set.
+        :param value: The value to set for the given key.
 
-    def __setitem__(self, key, value):
+        """
         if self.keyNormalizer is not None:
             key = self.keyNormalizer.__func__(key)
         if self.valueNormalizer is not None:
             value = self.valueNormalizer.__func__(value)
         self._setItem(key, value)
 
-    def _setItem(self, key, value):
-        """
-        Subclasses must override this method.
+    def _setItem(self, key: Any, value: Any) -> None:
+        """Set the value for a given key in the native object.
+
+        This is the environment implementation of :meth:`BaseDict.__setitem__`.
+
+        :param key: The key to set. If a :cvar:`BaseDict.keyNormalizer`
+            is set, it will have been applied to the given key.
+        :param value: The value to set for the given key. If
+            a :cvar:`BaseDict.valueNormalizer` is set, it will have been applied
+            to the given value.
+        :raises NotImplementedError: If the method has not been overridden by a
+            subclass.
+
+        .. important::
+
+            Subclasses must override this method.
+
         """
         self.raiseNotImplementedError()
 
-    def __getitem__(self, key):
+    def __getitem__(self, key: Any) -> Any:
+        """Get the value for a given key from the object.
+
+        :param key: The key to retrieve the value for.
+        :return: The value for the given key.
+
+        """
         if self.keyNormalizer is not None:
             key = self.keyNormalizer.__func__(key)
         value = self._getItem(key)
         if self.valueNormalizer is not None:
             value = self.valueNormalizer.__func__(value)
         return value
 
-    def _getItem(self, key):
-        """
-        Subclasses must override this method.
+    def _getItem(self, key: Any) -> Any:
+        """Get the value for a given key from the native object.
+
+        This is the environment implementation of :meth:`BaseDict.__getitem__`.
+
+        :param key: The key to retrieve the value for. If
+            a :cvar:`BaseDict.keyNormalizer` is set, it will have been applied
+            to the key.
+        :return: The value for the given key. If
+            a :cvar:`BaseDict.valueNormalizer` is set, it will have been applied
```

Comment:
better to say `it will be applied in the calling method to the returned value`

Reply:
Changed.